File: test/four/app/src/main/java/com/z3d/display/particle/CombineParticleData.java
package com.z3d.display.particle;

import android.util.Log;

import com.z3d.base.ByteArray;
import com.z3d.base.Scene_data;
import com.z3d.display.particle.ball.ParticleBallData;
import com.z3d.display.particle.facet.ParticleFacetData;
import com.z3d.display.particle.locus.ParticleLocusData;
import com.z3d.display.particle.locusball.ParticleLocusballData;
import com.z3d.display.particle.model.Display3DModelPartilce;
import com.z3d.display.particle.model.ParticleModelData;
import com.z3d.engine.ResCount;
import com.z3d.scene.Scene3D;

import java.util.ArrayList;
import java.util.List;

public class CombineParticleData extends ResCount {

    public  float maxTime;
    public  List<ParticleData> dataAry;

    public  CombineParticleData(Scene3D val  ){
  super(val);
    }
    public void setDataByte(ByteArray _byte) {
        _byte.byteBuffer.position=0;
        int version = _byte.readInt();
        int len = _byte.readInt();
        this.maxTime = 0;
        this.dataAry = new ArrayList();
        for (int i = 0; i < len; i++) {
            int particleType = _byte.readInt();
            ParticleData pdata  = this.getParticleDataType(particleType);
            pdata.version = version;
            pdata.setAllByteInfo(_byte);
            this.dataAry.add(pdata);
            if (pdata.timelineData.maxFrameNum > this.maxTime) {
                this.maxTime = pdata.timelineData.maxFrameNum;
            }
        }
        this.maxTime *= Scene_data.frameTime;
    }
    private ParticleData getParticleDataType(int type)
    {
        ParticleData pdata=null;
        switch (type){
            case 1:
                pdata=new ParticleFacetData(this.scene3d);
                break;
            case 3:
                pdata=new ParticleLocusData(this.scene3d);
                break;
            case 4:
            {
                pdata = new ParticleModelData(this.scene3d);
                break;
            }
            case 7:
            {
                pdata = new ParticleModelData(this.scene3d);
                break;
            }
            case 9:
            {
                pdata = new ParticleModelData(this.scene3d);
                break;
            }
            case 14:
            {
                pdata = new ParticleLocusballData(this.scene3d);
                break;
            }
            case 18:
            {
                pdata = new ParticleBallData(this.scene3d);
                break;
            }
            default:
                Log.d("没有对象", "getParticleDataType: ");
                break;
        }
        return pdata;
    }

    public CombineParticle getCombineParticle()
    {
        CombineParticle particle = new CombineParticle();
        particle.maxTime = this.maxTime;
        for (int i = 0; i < this.dataAry.size(); i++) {
            Display3DParticle display  = this.dataAry.get(i).creatPartilce();
            if (display instanceof Display3DModelPartilce) {

            }
            particle.addPrticleItem(display);

        }
        particle.sourceData = this;
        return particle;
    }

}

File: metalios/listios/StorageStarted/pan3d/display/particle/CombineParticleData.m
//
//  CombineParticleData.m
//  iosgl
//
//  Created by zhao on 23/2/2020.
//  Copyright © 2020 zhao. All rights reserved.
//
#import "Scene_data.h"
#import "CombineParticleData.h"
#import "ParticleData.h"
#import "ParticleFacetData.h"
#import "ParticleLocusData.h"
#import "ParticleBallData.h"
#import "ParticleFollowData.h"
#import "CombineParticle.h"
#import "ParticleLocusballData.h"
#import "ParticleModelData.h"
#import "Display3DParticle.h"

@implementation CombineParticleData
-(void)setDataByte:(ByteArray*)byte;
{
    int version=[byte readInt];
    int len=[byte readInt];
    self.maxTime=0;
    self.dataAry=[[NSMutableArray alloc]init];
    for (int i = 0; i < len; i++) {
        int particleType=[byte readInt];
        ParticleData *pdata= [self getParticleDataType:particleType];
        if(pdata){
            pdata.version=version;
            [pdata setAllByteInfo:byte];
            if ([pdata isKindOfClass:[ParticleModelData class]]) {
                if(self.dataAry.count==0){
                }
            }
           [self.dataAry addObject:pdata];
            if (pdata.timelineData.maxFrameNum > self.maxTime) {
                self.maxTime = pdata.timelineData.maxFrameNum;
            }
        }
        
    }
    self.maxTime *= [Scene_data default].frameTime;
 
}
-(ParticleData*)getParticleDataType:(int)type;
{
    ParticleData *pdata;
    switch (type) {
        case 1:
            pdata = [[ParticleFacetData alloc]init:self.scene3D];
            break;
        case 3:
            pdata = [[ParticleLocusData alloc]init:self.scene3D];
            break;
        case 8:
            pdata = [[ParticleFollowData alloc]init:self.scene3D];
            break;
        case 4:
            pdata = [[ParticleModelData alloc]init:self.scene3D];
            break;
        case 7:
            pdata = [[ParticleModelData alloc]init:self.scene3D];
            break;
        case 9:
            pdata = [[ParticleModelData alloc]init:self.scene3D];
            break;
        case 14:
            pdata = [[ParticleLocusballData alloc]init:self.scene3D];
            break;
        case 18:
            pdata = [[ParticleBallData alloc]init:self.scene3D];
            break;
        default:
            NSLog(@"没有的类型  %d",type);
            break;
    }
    
    return pdata;
}


-(CombineParticle*)getCombineParticle;
{
    CombineParticle* particle=[[CombineParticle alloc]init];
    particle.maxTime=self.maxTime;
    for (int i = 0; i < self.dataAry.count; i++) {
        Display3DParticle *display = [((ParticleData*)self.dataAry[i]) creatPartilce];
        [particle addPrticleItem:display];
    }
    particle.sourceData = self;
    return particle;
}
@end
